<!doctype html><html lang=en><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="IE=edge,chrome=1"><title>Exceptions - COMP2511 Musings</title><meta name=renderer content=webkit><meta name=viewport content="width=device-width,initial-scale=1,maximum-scale=1"><meta http-equiv=cache-control content=no-transform><meta http-equiv=cache-control content=no-siteapp><meta name=theme-color content=#f8f5ec><meta name=msapplication-navbutton-color content=#f8f5ec><meta name=apple-mobile-web-app-capable content=yes><meta name=apple-mobile-web-app-status-bar-style content=#f8f5ec><meta name=author content=z5206677><meta name=description content="Exceptions are useful to programmatically handle both foreseen, and (somewhat) unforeseen errors."><meta name=keywords content=Hugo,theme,even><meta name=generator content="Hugo 0.58.2 with theme even"><link rel=canonical href=../../post/exceptions/><link rel=apple-touch-icon sizes=180x180 href=../../apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=../../favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=../../favicon-16x16.png><link rel=manifest href=../../manifest.json><link rel=mask-icon href=../../safari-pinned-tab.svg color=#5bbad5><link href=../../sass/main.min.b5a744db6de49a86cadafb3b70f555ab443f83c307a483402259e94726b045ff.css rel=stylesheet><link rel=stylesheet href=https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.css integrity="sha256-7TyXnr2YU040zfSP+rEcz29ggW4j56/ujTPwjMzyqFY=" crossorigin=anonymous><link rel=stylesheet href=../../css/typedjs.shortcode.css><meta property=og:title content=Exceptions><meta property=og:description content="Exceptions are useful to programmatically handle both foreseen, and (somewhat) unforeseen errors."><meta property=og:type content=article><meta property=og:url content=/post/exceptions/><meta property=article:published_time content=2019-10-22T14:10:36+11:00><meta property=article:modified_time content=2019-10-22T14:10:36+11:00><meta itemprop=name content=Exceptions><meta itemprop=description content="Exceptions are useful to programmatically handle both foreseen, and (somewhat) unforeseen errors."><meta itemprop=datePublished content=2019-10-22T14:10:36&#43;11:00><meta itemprop=dateModified content=2019-10-22T14:10:36&#43;11:00><meta itemprop=wordCount content=358><meta itemprop=keywords content><meta name=twitter:card content=summary><meta name=twitter:title content=Exceptions><meta name=twitter:description content="Exceptions are useful to programmatically handle both foreseen, and (somewhat) unforeseen errors."><!--[if lte IE 9]><script src=https://cdnjs.cloudflare.com/ajax/libs/classlist/1.1.20170427/classList.min.js></script><![endif]--><!--[if lt IE 9]><script src=https://cdn.jsdelivr.net/npm/html5shiv@3.7.3/dist/html5shiv.min.js></script><script src=https://cdn.jsdelivr.net/npm/respond.js@1.4.2/dest/respond.min.js></script><![endif]--></head><body><div id=mobile-navbar class=mobile-navbar><div class=mobile-header-logo><a href=../../ class=logo>COMP2511 Musings</a></div><div class=mobile-navbar-icon><span></span><span></span><span></span></div></div><nav id=mobile-menu class="mobile-menu slideout-menu"><ul class=mobile-menu-list><a href=../../><li class=mobile-menu-item>Home</li></a><a href=https://github.com/featherbear/UNSW-COMP2511><li class=mobile-menu-item>GitHub</li></a></ul></nav><div class=container id=mobile-panel><header id=header class=header><div class=logo-wrapper><a href=../../ class=logo>COMP2511 Musings</a></div><nav class=site-navbar><ul id=menu class=menu><li class=menu-item><a class=menu-item-link href=../../>Home</a></li><li class=menu-item><a class=menu-item-link href=https://github.com/featherbear/UNSW-COMP2511>GitHub</a></li></ul></nav></header><main id=main class=main><div class=content-wrapper><div id=content class=content><article class=post><header class=post-header><h1 class=post-title>Exceptions</h1><div class=post-meta><span class=post-time>2019-10-22</span></div></header><div class=post-toc id=post-toc><h2 class=post-toc-title>Contents</h2><div class="post-toc-content always-active"><nav id=TableOfContents><ul><li><a href=#throwing-exceptions>Throwing Exceptions</a></li><li><a href=#everything-is-an-exception>Everything is an <code>Exception</code></a></li><li><a href=#defining-custom-exceptions>Defining Custom Exceptions</a></li><li><a href=#delegating-raised-exceptions-to-other-code>Delegating raised exceptions to other code</a></li><li><a href=#exceptions-and-inheritance>Exceptions and Inheritance</a></li></ul></nav></div></div><div class=post-content><p>Exceptions are useful to programmatically handle both foreseen, and (somewhat) unforeseen errors.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span><span class=lnt>6
</span><span class=lnt>7
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=k>try</span> <span class=p>{</span>
  <span class=n>doSomething</span><span class=p>()</span>
<span class=p>}</span> <span class=k>catch</span> <span class=p>(</span><span class=n>ExceptionThing</span> <span class=nf>err</span><span class=p>)</span> <span class=p>{</span>
  <span class=n>System</span><span class=p>.</span><span class=na>out</span><span class=p>.</span><span class=na>println</span><span class=p>(</span><span class=s>&#34;Nope!&#34;</span><span class=p>);</span>
<span class=p>}</span> <span class=k>finally</span> <span class=p>{</span>
  <span class=n>System</span><span class=p>.</span><span class=na>out</span><span class=p>.</span><span class=na>println</span><span class=p>(</span><span class=s>&#34;Anyways...&#34;</span><span class=p>);</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><h1 id=throwing-exceptions>Throwing Exceptions</h1><p><code>throw new SomeExceptionClass(...)</code></p><h1 id=everything-is-an-exception>Everything is an <code>Exception</code></h1><p>That is, <code>Exception</code> is the base class of all exceptions.</p><p>We could technically catch every single possible exception with <code>... } catch (Exception e) { ... }</code></p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=k>try</span> <span class=p>{</span>
  <span class=n>doSomething</span><span class=p>();</span>
<span class=p>}</span> <span class=k>catch</span> <span class=p>(</span><span class=n>Exception</span> <span class=nf>e</span><span class=p>)</span> <span class=p>{</span>
  <span class=c1>// pass
</span><span class=c1></span><span class=p>}</span></code></pre></td></tr></table></div></div><p>Note: This is discouraged, as you lose scope of which exception was raised.</p><h1 id=defining-custom-exceptions>Defining Custom Exceptions</h1><p>To define a custom Java exception, we need to extend the abstract <code>Exception</code> class</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kd>public</span> <span class=nf>class</span> <span class=n>MyException</span> <span class=nf>extends</span> <span class=n>Exception</span> <span class=p>{</span>
    <span class=kd>private</span> <span class=nf>String</span> <span class=n>message</span><span class=p>;</span>

    <span class=kd>public</span> <span class=nf>MyException</span><span class=p>(</span><span class=n>String</span> <span class=nf>message</span><span class=p>)</span> <span class=p>{</span>
      <span class=k>this</span><span class=p>.</span><span class=na>message</span> <span class=o>=</span> <span class=n>message</span><span class=p>;</span>
    <span class=p>}</span>

    <span class=kd>public</span> <span class=nf>String</span> <span class=n>getMessage</span><span class=p>()</span> <span class=p>{</span>
      <span class=k>return</span> <span class=k>this</span><span class=p>.</span><span class=na>message</span>
    <span class=p>}</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><p>We can then throw, and catch this exception.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt> 1
</span><span class=lnt> 2
</span><span class=lnt> 3
</span><span class=lnt> 4
</span><span class=lnt> 5
</span><span class=lnt> 6
</span><span class=lnt> 7
</span><span class=lnt> 8
</span><span class=lnt> 9
</span><span class=lnt>10
</span><span class=lnt>11
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kt>void</span> <span class=nf>doThing</span><span class=p>()</span> <span class=p>{</span>
  <span class=k>throw</span> <span class=k>new</span> <span class=n>MyException</span><span class=p>(</span><span class=s>&#34;Uh oh!&#34;</span><span class=p>);</span>
<span class=p>}</span>

<span class=kt>void</span> <span class=nf>doOtherThing</span><span class=p>()</span> <span class=p>{</span>
  <span class=k>try</span> <span class=p>{</span>
    <span class=n>doOtherThing</span><span class=p>();</span>
  <span class=p>}</span> <span class=k>catch</span> <span class=p>(</span><span class=n>MyException</span> <span class=nf>e</span><span class=p>)</span> <span class=p>{</span>
    <span class=n>System</span><span class=p>.</span><span class=na>out</span><span class=p>.</span><span class=na>println</span><span class=p>(</span><span class=n>e</span><span class=p>.</span><span class=na>getMessage</span><span class=p>());</span>
  <span class=p>}</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><h1 id=delegating-raised-exceptions-to-other-code>Delegating raised exceptions to other code</h1><p>Consider the function</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kt>void</span> <span class=nf>tryDoIt</span><span class=p>()</span> <span class=p>{</span>
  <span class=n>OpenFile</span><span class=p>(</span><span class=s>&#34;test.txt&#34;</span><span class=p>);</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><p>If <code>OpenFile</code> has the possibility to throw an exception (i.e. <code>FileNotFoundException</code>), then we will have to acknowledge that possibility in our code</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span><span class=lnt>6
</span><span class=lnt>7
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kt>void</span> <span class=nf>tryDoIt</span><span class=p>()</span> <span class=p>{</span>
  <span class=k>try</span> <span class=p>{</span>
    <span class=n>OpenFile</span><span class=p>(</span><span class=s>&#34;test.txt&#34;</span><span class=p>);</span>
  <span class=p>}</span> <span class=k>catch</span> <span class=p>(</span><span class=n>FileNotFoundException</span> <span class=nf>e</span><span class=p>)</span> <span class=p>{</span>
    <span class=n>System</span><span class=p>.</span><span class=na>out</span><span class=p>.</span><span class=na>println</span><span class=p>(</span><span class=s>&#34;File not found!!!&#34;</span><span class=p>);</span>
  <span class=p>}</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><p>If we wanted to catch the exception <strong>outside</strong> of our <code>tryDoIt</code> function, we would have to reraise the exception.</p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span><span class=lnt>4
</span><span class=lnt>5
</span><span class=lnt>6
</span><span class=lnt>7
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kt>void</span> <span class=nf>tryDoIt</span><span class=p>()</span> <span class=p>{</span>
  <span class=k>try</span> <span class=p>{</span>
    <span class=n>OpenFile</span><span class=p>(</span><span class=s>&#34;test.txt&#34;</span><span class=p>);</span>
  <span class=p>}</span> <span class=k>catch</span> <span class=p>(</span><span class=n>FileNotFoundException</span> <span class=nf>e</span><span class=p>)</span> <span class=p>{</span>
    <span class=k>throw</span> <span class=n>e</span><span class=p>;</span>
  <span class=p>}</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><p>However in Java, we can add a hint to the method definition that it may throw the <code>FileNotFoundException</code></p><div class=highlight><div class=chroma><table class=lntable><tr><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=lnt>1
</span><span class=lnt>2
</span><span class=lnt>3
</span></code></pre></td><td class=lntd><pre class=chroma><code class=language-java data-lang=java><span class=kt>void</span> <span class=nf>tryDoIt</span><span class=p>()</span> <span class=kd>throws</span> <span class=nf>FileNotFoundException</span> <span class=p>{</span>
  <span class=n>OpenFile</span><span class=p>(</span><span class=s>&#34;test.txt&#34;</span><span class=p>);</span>
<span class=p>}</span></code></pre></td></tr></table></div></div><p>By using the <code>throws</code> keyword, we can delegate the handling of the exception to another function.</p><h1 id=exceptions-and-inheritance>Exceptions and Inheritance</h1><p>Inherited classes must not throw more exceptions than those defined by their parent - only subsets.</p></div><footer class=post-footer><nav class=post-nav><a class=prev href=../../post/composite-pattern/><i class="iconfont icon-left"></i><span class="prev-text nav-default">Composite Pattern</span>
<span class="prev-text nav-mobile">Prev</span></a>
<a class=next href=../../post/assign2-stories-1/><span class="next-text nav-default">Assignment Two - User Stories (rev 1)</span>
<span class="next-text nav-mobile">Next</span>
<i class="iconfont icon-right"></i></a></nav></footer></article><script>(function(f,a,t,h,o,m){a[h]=a[h]||function(){(a[h].q=a[h].q||[]).push(arguments)};o=f.createElement('script'),m=f.getElementsByTagName('script')[0];o.async=1;o.src=t;o.id='fathom-script';m.parentNode.insertBefore(o,m)})(document,window,'//ss.featherbear.cc/tracker.js','fathom');fathom('set','siteId','NEQTU');fathom('trackPageview');</script></div></div></main><footer id=footer class=footer><div class=social-links><a href=mailto:z5206677@student.unsw.edu.au class="iconfont icon-email" title=email></a><a href=https://www.linkedin.com/in/andrewjinmengwong/ class="iconfont icon-linkedin" title=linkedin></a><a href=https://github.com/featherbear class="iconfont icon-github" title=github></a><a href=https://www.instagram.com/_andrewjwong/ class="iconfont icon-instagram" title=instagram></a><a href=../../index.xml type=application/rss+xml class="iconfont icon-rss" title=rss></a></div><div class=copyright><span class=power-by>Powered by <a class=hexo-link href=https://gohugo.io>Hugo</a></span>
<span class=division>|</span>
<span class=theme-info>Theme -
<a class=theme-link href=https://github.com/olOwOlo/hugo-theme-even>Even</a></span>
<span class=copyright-year>&copy;
2019 -
2021<span class=heart><i class="iconfont icon-heart"></i></span><span>Andrew Wong (z5206677)</span></span></div></footer><div class=back-to-top id=back-to-top><i class="iconfont icon-up"></i></div></div><script src=https://cdn.jsdelivr.net/npm/jquery@3.2.1/dist/jquery.min.js integrity="sha256-hwg4gsxgFZhOsEEamdOYGBf13FyQuiTwlAQgxVSNgt4=" crossorigin=anonymous></script><script src=https://cdn.jsdelivr.net/npm/slideout@1.0.1/dist/slideout.min.js integrity="sha256-t+zJ/g8/KXIJMjSVQdnibt4dlaDxc9zXr/9oNPeWqdg=" crossorigin=anonymous></script><script src=https://cdn.jsdelivr.net/npm/@fancyapps/fancybox@3.1.20/dist/jquery.fancybox.min.js integrity="sha256-XVLffZaxoWfGUEbdzuLi7pwaUJv1cecsQJQqGLe7axY=" crossorigin=anonymous></script><script type=text/javascript src=../../js/main.min.c12618f9a600c40bd024996677e951e64d3487006775aeb22e200c990006c5c7.js></script><script type=application/javascript>var doNotTrack=false;if(!doNotTrack){window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;ga('create','UA-107434487-2','auto');ga('send','pageview');}</script><script async src=https://www.google-analytics.com/analytics.js></script><script src=../../js/typed.js@2.0.9></script><script src=../../js/typedjs.shortcode.js></script></body></html>